### PR TITLE
#1527 fixed

### LINKED ISSUE #1158: Observed data export crash
@Yuri05 commented on [Thu Jul 16 2020](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/1527)

[ExportCrash.zip](https://github.com/Open-Systems-Pharmacology/PK-Sim/files/4932235/ExportCrash.zip)
When exporting observed data to Excel:

```
Application:
PK-Sim® 9.0.170

InvalidSheetName

Stack trace:
at l7.d(String A_0)
   at l7.b(String A_0)
   at l7.e(String A_0)
   at re.a(Int32 A_0, String A_1)
   at OSPSuite.Infrastructure.Export.ExportToExcelTask.exportDataTableToWorkBook(WorkBook workBook, DataTable dataTable, Int32 sheetIndex)
   at OSPSuite.Infrastructure.Export.ExportToExcelTask.ExportDataTablesToExcel(IEnumerable`1 dataTables, String fileName, Boolean openExcel, Action`2 workbookConfiguration)
   at OSPSuite.Infrastructure.Export.ExportToExcelTask.ExportDataTablesToExcel(IEnumerable`1 dataTables, String fileName, Boolean openExcel, Action`1 workbookConfiguration)
   at OSPSuite.Infrastructure.Export.DataRepositoryExportTask.ExportToExcel(IEnumerable`1 dataTables, String fileName, Boolean launchExcel)
   at OSPSuite.Infrastructure.Export.DataRepositoryExportTask.ExportToExcel(IEnumerable`1 dataColumns, String fileName, Boolean launchExcel, DataColumnExportOptions exportOptions)
   at OSPSuite.Core.Domain.Services.ObservedDataTask.Export(DataRepository observedData)
   at OSPSuite.Presentation.UICommands.ExportObservedDataToExcelCommand.PerformExecute()
   at OSPSuite.Presentation.UICommands.ObjectUICommand`1.Execute()
   at OSPSuite.Presentation.UICommands.UICommandExtensions.ExecuteWithinExceptionHandler(IUICommand uiCommand, IEventPublisher eventPublisher, IChangePropagator changePropagator)
```

---

@Yuri05 commented on [Thu Jul 16 2020](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/1527#issuecomment-659468426)

@Incei FYI

---

@msevestre commented on [Thu Jul 16 2020](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/1527#issuecomment-659469822)

duplicate of https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/1494 ?

---

@Yuri05 commented on [Thu Jul 16 2020](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/1527#issuecomment-659473273)

could be. At least reproducible now :)

---

@Yuri05 commented on [Mon Jul 05 2021](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/1527#issuecomment-873946051)

@abdelr @georgeDaskalakis Just tried it with PK-Sim 10: still crashes, now with the error message
```
Application:
PK-Sim® 10.0.170

Invalid sheet name 'Digoxin, 0.5 mg colon, Ochs ''. Sheet names must not begin or end with (').

Stack trace:

at NPOI.SS.Util.WorkbookUtil.ValidateSheetName(String sheetName)
   at NPOI.XSSF.UserModel.XSSFWorkbook.CreateSheet(String sheetname)
   at OSPSuite.Infrastructure.Export.ExportToExcelTask.exportDataTableToWorkBook(IWorkbookConfiguration workBookConfiguration, DataTable dataTable)
   at OSPSuite.Infrastructure.Export.ExportToExcelTask.ExportDataTablesToExcel(IEnumerable`1 dataTables, String fileName, Boolean openExcel, IWorkbookConfiguration workBookConfiguration)
   at OSPSuite.Infrastructure.Export.DataRepositoryExportTask.ExportToExcel(IEnumerable`1 dataTables, String fileName, Boolean launchExcel)
   at OSPSuite.Infrastructure.Export.DataRepositoryExportTask.ExportToExcel(IEnumerable`1 dataColumns, String fileName, Boolean launchExcel, DataColumnExportOptions exportOptions)
   at OSPSuite.Core.Domain.Services.ObservedDataTask.Export(DataRepository observedData)
   at OSPSuite.Presentation.UICommands.ExportObservedDataToExcelCommand.PerformExecute()
   at OSPSuite.Presentation.UICommands.ObjectUICommand`1.Execute()
   at OSPSuite.Presentation.UICommands.UICommandExtensions.ExecuteWithinExceptionHandler(IUICommand uiCommand, IEventPublisher eventPublisher, IChangePropagator changePropagator)
```

---

@georgeDaskalakis commented on [Mon Jul 05 2021](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/1527#issuecomment-874302341)

The problem here is the single quote character ( ' ) in the name of the observed data set in exactly the 30th position. The maximum allowed name dor the sheet is 29 characters, then the name of the sheet ends up having a ' at the end, which is not accepted by excel. I have added a cleaning up of any ' characters left at the beginning or in the end of the name ater the truncation.

### PR DESCRIPTION
**Pull request by georgeDaskalakis**
closes #1158

### CODE REVIEW

File: src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs [hunk 1]
```diff
@@ -108,8 +108,15 @@ private static string tableNameFrom(DataColumn baseGrid)
          var tableName = baseGrid.Repository != null ? baseGrid.Repository.Name : defaultTableName;
          if (string.IsNullOrEmpty(tableName))
             tableName = defaultTableName;
```
Comment (Yuri05) [inline src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:110]: could replace line 110 with 
`return defaultTableName;`

File: src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs [hunk 2]
```diff
@@ -108,8 +108,15 @@ private static string tableNameFrom(DataColumn baseGrid)
...
+         var correctedTableName = new string(FileHelper.RemoveIllegalCharactersFrom(tableName).Take(Constants.MAX_NUMBER_OF_CHAR_IN_TABLE_NAME).ToArray());
```
Comment (Yuri05) [inline src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:111]: 2 remarks:
* `[` and `]` are two chars which are not allowed in the excel sheet names but are allowed in the file names, so I guess `FileHelper.RemoveIllegalCharacters` will not remove them
* Why is `.ToArray()` required?

File: src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs [hunk 3]
```diff
@@ -108,8 +108,15 @@ private static string tableNameFrom(DataColumn baseGrid)
...
 
-         return new string(FileHelper.RemoveIllegalCharactersFrom(tableName).Take(Constants.MAX_NUMBER_OF_CHAR_IN_TABLE_NAME).ToArray());
+         while (correctedTableName.EndsWith("'") && correctedTableName.Length > 0)
```
Comment (Yuri05) [inline src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:113]: if `correctedTableName.EndsWith("'")` is **true** then `correctedTableName.Length` cannot be 0, thus the 2nd condition can be removed, right?

File: src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs [hunk 4]
```diff
@@ -108,8 +108,15 @@ private static string tableNameFrom(DataColumn baseGrid)
...
+            correctedTableName = correctedTableName.Substring(0, correctedTableName.Length - 1);
+
+         while (correctedTableName.StartsWith("'") && correctedTableName.Length > 0)
```
Comment (Yuri05) [inline src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:116]: same here

File: src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs [hunk 5]
```diff
@@ -108,8 +108,15 @@ private static string tableNameFrom(DataColumn baseGrid)
...
+            correctedTableName = correctedTableName.Substring(1);
+
+         return correctedTableName;
```
Comment (Yuri05) [inline src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:119]: weird use case, but if the original name contains only` '`, then correctedTableName will be empty at the end :)

File: src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs (hunk 1 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:110]: done

File: src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs (hunk 3 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:113]: true. corrected

File: src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs (hunk 4 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:116]: also corrected

File: src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs (hunk 5 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:119]: yes the point of the test was to make sure that we do not go out ouf bounds in traversing the string.  Empty string is OK: later on the export task replaces the empty names with "Sheet1" etc.

Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:119]: there is even this test: When_converting_a_data_repository_with_only_single_quotes for exactly that use case

Comment (Yuri05) [reply src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:119]: ok, cool

File: src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs (hunk 2 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:111]: To be honest this line was the same before: I have just created the correctedTableName variable to remove the ' before returning it. The way it is written, it needs to be an array for new string() to work. Do you think this line should be somehow refactored?

About the [ and ] characters: I added functionality removing them (also a test for that).

Comment (Yuri05) [reply src/OSPSuite.Infrastructure.Export/DataRepositoryExportTask.cs:113]: >Do you think this line should be somehow refactored?

Just checked, that Take returns IEnumerable<char> and not a string, that's why. Is ok then.

**Review by Yuri05 — APPROVED** [review]

**Pull request merged**